src/pterms/PtStore.cc

```diff
@@ -67,49 +99,88 @@ SymRef PtStore::lookupSymbol(const char* s, const vec<PTRef>& args) {
                     int j = 0;
                     for (; j < args.size(); j++) {
                         SymRef argt = pta[args[j]].symb();
                         if (symstore[argt].rsort() != t[0]) break;
                     }
-                    if (j == args.size()) return ctr;
-                }
-                else if (t.nargs() < args.size_() && t.pairwise()) {
+                    if (j == args.size()) {
+                        candidates.push(ctr);
+                    }
+                } else if (t.nargs() < args.size_() && t.pairwise()) {
                     int j = 0;
                     for (; j < args.size(); j++) {
                         SymRef argt = pta[args[j]].symb();
                         if (symstore[argt].rsort() != t[0]) break;
                     }
-                    if (j == args.size()) return ctr;
+                    if (j == args.size()) {
+                        candidates.push(ctr);
+                    }
                 }
             }
-        }
-
-        for (int i = 0; i < trefs.size(); i++) {
-            SymRef ctr = trefs[i];
-            const Symbol& t = symstore[ctr];
-            if (t.nargs() == args.size_()) {
-                // t is a potential match.  Check that arguments match
-                uint32_t j = 0;
-                for (; j < t.nargs(); j++) {
-                    SymRef argt = pta[args[j]].symb();
-                    if (t[j] != symstore[argt].rsort()) break;
+        } else {
+            for (int i = 0; i < trefs.size(); i++) {
+                SymRef ctr = trefs[i];
+                const Symbol &t = symstore[ctr];
+                if (t.nargs() == args.size_()) {
+                    // t is a potential match.  Check that arguments match
+                    uint32_t j = 0;
+                    for (; j < t.nargs(); j++) {
+                        SymRef argt = pta[args[j]].symb();
+                        if (t[j] != symstore[argt].rsort()) break;
+                    }
+                    if (j == t.nargs()) {
+                        candidates.push(ctr);
+                    }
                 }
-                if (j == t.nargs())
-                    return ctr;
             }
         }
+    }
+    if (candidates.size() == 0) {
+        return SymRef_Undef; // Not found
+    } else if (candidates.size() == 1 and sort == SRef_Undef) {
+        return candidates[0];
+    } else if (candidates.size() == 1 and sort != SRef_Undef) {
+        if (symstore[candidates[0]].rsort() != sort) {
+            return SymRef_Undef;
+        } else {
+            return candidates[0];
+        }
+    }
 
+    assert(candidates.size() > 1);
+    if (sort == SRef_Undef) {
+        throw OsmtApiException("Ambiguous symbol: `" + std::string(s) + "'");
+    }
+
+    assert(sort != SRef_Undef);
+
+    vec<SymRef> retSortMatched;
+    retSortMatched.capacity(candidates.size());
+    for (SymRef sr : candidates) {
```

Comment:
I think you can use `std::count_if` here and avoid the temporary vector. I think it will simplified the code a bit as well.

Reply:
I tried to come up with something simpler.